eclipse-theia/theia
# [Debug view] Add dynamic debug configurations

#### What it does


Following the support for dynamic debug configurations via
command Debug: Select and Start debugging. PR #10134

This change extends the functionality further to bring provided
dynamic debug configurations to the list of available configurations
under the debug view.

The user can select these dynamic configurations by first selecting
the debug type and a quick pick input box will follow to allow the
selection from the available configurations of the selected type.

Once a dynamic configuration has been selected it will be added to a
short list (limited to 3) of recently used dynamic debug
configurations, these entries are available to the user, rendered
with the name and suffixed with the debug type in parenthesis.
This will facilitate subsequent selection / execution.

This change additionally preserves and restores the list of recently
selected  dynamic debug configurations so they are presented in
subsequent sessions.

These configurations are refreshed from the providers when:
   - Configuration providers are registered or unregistered
   - Focus is gained or lost by the configuration selection box

Refreshing these configurations intends to render valid dynamic
configurations to the current context.
e.g.
   - Honoring an extension 'when' clause to a file with a specific
     extension opened in the active editor.

However there are situations where the context for execution of a
dynamic configuration may no longer be valid e.g.
   - The configuration is restored from a previous session and the
     currently selected file is not supported.
   - The switch of context may not have involved a refresh of dynamic
     debug configurations. (e.g. switching active editors, etc.)
Considering the above, execution of dynamic configurations triggers
the fetch of dynamic configurations for the selected provider type,
if the configuration is no longer provided, the user will be notified
of a missing configuration or not applicable to the current context.

It additionally
Fixes: #11050
The update were required to integrate  the solution with the `SelectComponent`



#### How to test

1. Assuming the extension `ms-vscode.js-debug` is installed as a builtin dependency
Open a `javascript` or `typescript` project, build it  and open in editor a `.js` or `.ts` file ready to be debugged.
`Open the Debug view and open the selection box.`
2. Select the entry with an ellipsis suffix (...) after a separator e.g.  `pwa-node...`
3. A quick pick will appear with the dynamic configurations provided, Select the one called `Run current file` and start it.
4. Verify that the debug session starts and that debug options work (e.g. stop at break points, etc.).
5. End the debug session,
6. Test that other non dynamic configurations are working as usual.
7. Verify that after selecting a dynamic configuration it becomes directly available under the main selection box
8. Restart the application and verify that the recently used dynamic configuration become available directly under the selection box.
9. Verify that a dynamic configuration selected in the main box can be restored after restarting the application.
10. Repeat the tests above using a `multi-root` workspace.

#To test with additional dynamic configurations, you can install a modified 'vscode-mock-debug' extension updated
to test `Theia`:
https://github.com/alvsan09/vscode-mock-debug/releases/download/v1.18-alpha/mock-debug-theia-0.46.3.vsix

The 'vscode-mock-debug' extension is meant to simulate debug on `.md` files although you can also try it on other types.
![dynamic_config_view gif](https://user-images.githubusercontent.com/76971376/138964851-4ecf6daf-5a70-4a1a-9c3e-8d877ecdc188.gif)


#### Review checklist


#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
